### DISCUSSION TITLE
An alternative to AutoMapper

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#91] [General]
# An alternative to AutoMapper

I am not the biggest fan of AutoMapper. It starts with good intentions but often ends up being a big mess. I have seen it used in many projects, and the configuration of the mappings is often scattered all over the place, and or they are huge!

https://steven-giesel.com/blogPost/3368bf96-648e-4fa2-9c6c-be71e40ac4cc

**Comment by darjanbogdan** [comment]
Indeed source generators are pretty good replacement for runtime mappers, however in the era of contextualized Copilot help, manual mapping is a single "tab" away. 

I'm wondering if avoiding boilerplate code will stop being a "thing" if it can be promptly generated.

**Reply by linkdotnet** [reply]
That is a good question. For sure, depending on the context, you can greatly reduce boilerplate code.
For example, in TypeScript, much of this mapping logic comes for "free".

**Comment by y2k4life** [comment]
I know AI and Copilots are the new shinny thing. But solutions have been solved without them.

I personally use MappingGenerator which generates codes without the need of AI, Copilots, or source generators.

**Reply by linkdotnet** [reply]
Nice approach. To what extend can you configure the tool?
Let’s say I want to flatten a list in a certain way?

**Reply by y2k4life** [reply]
Yes, this is a freemium product, and that feature is not available in the freemium version. The price is affordable $80.

>Allow for type flattening
>This option affects the source-to-target matching algorithm by extending source searching on nested properties when there is no direct match. It allows achieving mappings that project complex objects to a flat structure.

In simple terms it works by you creating a method signature with the input parameters and a return type. Right mouse click on the method then choose generate code.

```c#
public static UserDTO Map(UserEntity entity)
{
}
```

This is what is generated.


```c#
public static UserDTO Map(UserEntity entity)
{
    return new UserDTO
    {
        Id = entity.Id,
        Version = entity.Version,
        // INFO: Properties from the 'Name' sub-object matched directly to the main object properties
        FirstName = entity.Name.FirstName,
        LastName = entity.Name.LastName,
        // INFO: Properties from the 'MainAddress' sub-object matched directly to the main object properties
        City = entity.MainAddress.City,
        ZipCode = entity.MainAddress.ZipCode,
        Street = entity.MainAddress.Street,
        BuildingNo = entity.MainAddress.Building_No,
        FlatNo = entity.MainAddress.FlatNo
    };
}
```

**Comment by simonb65** [comment]
An easier option is to not use a mapper. It's a bad pattern.
If you do need a mapper use common interfaces so no mapping is required, or write the code.
Either way you need to do tests.

Good friends don't lets friends use mappers...

**Reply by linkdotnet** [reply]
> Good friends don't lets friends use mappers

Haha nice one. I am absolutely on your side - I, myself, prefer just "code".
But I do know teams that like tools like AutoMapper. And transitioning to something that is "compile-time constant" is still better in my opinion.

**Comment by PrashantUnity** [comment]
Hi this out of context question. 
If Possible Could you please tell me 
how did you implemented github comment in website (Blazor server application)

**Reply by linkdotnet** [reply]
I am using a service called [Giscus](https://github.com/giscus/giscus).
It is implemented in my blog [here](https://github.com/linkdotnet/Blog/blob/master/src/LinkDotNet.Blog.Web/Features/ShowBlogPost/Components/Giscus.razor).

I hope that helps!
